pkg/webhook/resources/upgrade/validator.go

```diff
@@ -252,34 +255,33 @@ func (v *upgradeValidator) checkDiskSpace(node *corev1.Node, minFreeDiskSpace ui
 	}
 
 	kubeletPort := node.Status.DaemonEndpoints.KubeletEndpoint.Port
-	kubeletURL := fmt.Sprintf("https://%s:%d/stats/summary", internalIP, kubeletPort)
-	req, err := http.NewRequest("GET", kubeletURL, nil)
+	kubeletURL := fmt.Sprintf("https://%s:%d", internalIP, kubeletPort)
+	summary, err := v.getKubeletStatsSummary(node.Name, kubeletURL)
 	if err != nil {
-		return werror.NewInternalError(fmt.Sprintf("node %s, can't make http.NewRequest to get %s, err: %+v", node.Name, kubeletURL, err))
-	}
-	req.Header.Set("Authorization", "Bearer "+v.bearToken)
-
-	resp, err := v.httpClient.Do(req)
-	if err != nil {
-		return werror.NewInternalError(fmt.Sprintf("node %s, can't make http request to get %s, err: %+v", node.Name, kubeletURL, err))
+		return err
 	}
-	defer resp.Body.Close()
 
-	if resp.StatusCode != http.StatusOK {
-		return werror.NewInternalError(fmt.Sprintf("node %s, http response from %s is %d", node.Name, kubeletURL, resp.StatusCode))
+	if summary.Node.Fs == nil || summary.Node.Fs.AvailableBytes == nil || summary.Node.Fs.CapacityBytes == nil || summary.Node.Fs.UsedBytes == nil {
+		return werror.NewInternalError(fmt.Sprintf("can't get node %s filesystem stats from %s, err: %+v", node.Name, kubeletURL, err))
 	}
 
-	var summary kubeletv1.Summary
-	body, err := ioutil.ReadAll(resp.Body)
+	kubeletConfiguration, err := v.getKubeletConfigz(node.Name, kubeletURL)
 	if err != nil {
-		return werror.NewInternalError(fmt.Sprintf("node %s, can't read response from %s, err: %+v", node.Name, kubeletURL, err))
+		return err
 	}
-	if err = json.Unmarshal(body, &summary); err != nil {
-		return werror.NewInternalError(fmt.Sprintf("node %s, can't parse json response %s, err: %+v", node.Name, string(body), err))
+
+	imageGCHighThresholdPercent := defaultImageGCHighThresholdPercent
+	if kubeletConfiguration.ImageGCHighThresholdPercent != nil {
+		imageGCHighThresholdPercent = float64(*kubeletConfiguration.ImageGCHighThresholdPercent)
 	}
-	if summary.Node.Fs.AvailableBytes == nil {
-		return werror.NewInternalError(fmt.Sprintf("can't get node %s available bytes from %s, err: %+v", node.Name, kubeletURL, err))
+	usedPercent := (float64(*summary.Node.Fs.UsedBytes) / float64(*summary.Node.Fs.CapacityBytes)) * 100.0
+	logrus.Debugf("node %s uses %3f%% storage space, kubelet image garbage collection threshold is %3f%%", node.Name, usedPercent, imageGCHighThresholdPercent)
+
+	if usedPercent > imageGCHighThresholdPercent {
```

Comment:
@FrankYang0529 I suggest we use the condition `"userPercent + (expected new image capacity)" > imageGCHighThresholdPercent`. The problem we saw before is that after/during preloading images, a GC removes new images. We can add a reasonable value of `(expected new image capacity)` by doing some experience...

Reply:
Updated it. Thank you.